### PR TITLE
Closure conversion design

### PR DESCRIPTION
**Pull request by agocke**
Initial work on closure conversion design document.

### CODE REVIEW

File: docs/compilers/Design/Closure Conversion.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    return env.x + 1;
+}
+
+struct Environment
+{
+    int x;
+}
+```
+
+Instead of referring to local variables, the rewritten closure now references fields on the environment. A `struct` environment is used to prevent extra allocations, since `struct`s are normally allocated on the stack. However, the astute reader may notice a problem with this conversion -- `struct`s are always copied when passed as arguments to a function call! This means that writes to the environment field will not always propogate back to the local variable in the method `M`. To work around this flaw, all Environment variables are passed as `ref` arguments to the rewritten closures.
```
Comment (jcouv) [inline docs/compilers/Design/Closure Conversion.md]: Why not include the `ref` in the code sample above? `<>__Local(ref env);` and such.

File: docs/compilers/Design/Closure Conversion.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+{
+    int x = 0;
+    var positive = e.Where(n => n > 0);
+    ...
+}
+```
+
+In this case we're passing a closure to the `IEnumerable<int>.Where` function, which is expecting a delegate of type `Func<int, bool>`. Note that that delegate type is immutable -- it is defined in external code and cannot be changed. Therefore, rewriting external callsites to take an extra `Environment` argument is impossible. We have to choose a different strategy for acquiring the environment.
+
+What we do is use a `class` Environment for cases like this. With a `class` Environment, the previous environment can be rewritten as follows:
```
Comment (jcouv) [inline docs/compilers/Design/Closure Conversion.md]: the previous ~~environment~~ *program* can be ...?

File: docs/compilers/Design/Closure Conversion.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+## Optimization
+
+The final passes are optimization passes. They attempt to simplify the tree by removing intermediate scopes with no captured variables and otherwise correcting the tree to create the most efficient rewritten form. Optimization opportunities could include running escape analysis to determine if capture hoisting could be done via copy or done in a narrower scope.
+
+## Rewriting
+
+The rewriting phase simply walks the bound tree and, at each new scope, checks to see if the AnalysisResult contains a scope with rewriting information to process. If so, locals and closures are replaced with the substitutions provided by the tree. Practically, this means that the tree contains:
+
+1. A list of synthesized locals to add to each scope.
+
```
Comment (jcouv) [inline docs/compilers/Design/Closure Conversion.md]: Nit: extra lines un-necessary

**Comment by VSadov** [comment]
Looks good. 
Perhaps could have a bit more details about generic type parameters and handling of `this` and `base`.

Also we have a number of optimizations that try to cache allocated delegates in static fields, instance fields, locals. I wonder if it is worth calling out how these are going to be handled.

It seems, that is can be considered just a part of rewriting strategy as we have it now. However it might be better to do the corresponding analysis in the preceding stages and keep rewriting relatively dumb.

**Comment by agocke** [comment]
@VSadov Right, one of the main purposes is to keep the rewriting as dumb as possible. I'll follow up on this after merging.

**Comment by agocke** [comment]
Also, we have the whole LocalFunctionReferenceRewriting phase because too much logic is in the current rewriting.

**Pull request merged**